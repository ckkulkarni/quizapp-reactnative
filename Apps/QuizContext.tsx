import React from 'react';

export const QuizContext = React.createContext<{
  score: number;
  setScore: React.Dispatch<React.SetStateAction<number>>;
  name: string;
  setName: React.Dispatch<React.SetStateAction<string>>;
  email: string;
  setMail: React.Dispatch<React.SetStateAction<string>>;
  phone: string;
  setPhone: React.Dispatch<React.SetStateAction<string>>;
  selectedValue: string;
  setSelected: React.Dispatch<React.SetStateAction<string>>;
  trackScore: boolean;
  setTracker: React.Dispatch<React.SetStateAction<boolean>>;
}>({
  score: 0,
  setScore: () => {},
  name: '',
  setName: () => {},
  email: '',
  setMail: () => {},
  phone: '',
  setPhone: () => {},
  selectedValue: '',
  setSelected: () => {},
  trackScore: false,
  setTracker: () => {},
});
