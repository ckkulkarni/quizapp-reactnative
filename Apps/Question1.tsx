import React, {useContext, useState} from 'react';
import {
  Text,
  View,
  Button,
  Pressable,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack/lib/typescript/src/types';
import {QuizContext} from './QuizContext';
export default function Question1() {
  const {trackScore, score, setScore, selectedValue} = useContext(QuizContext);
  const navigation = useNavigation<NativeStackNavigationProp<any>>();
  const [selectedIndex, setIndex] = useState<number>();
  const currentScore = score;
  const maxScore = 9;
  const [submitted, setSubmit] = useState<boolean>(false);
  let questionText = '';
  let options: any[] = [];
  if (selectedValue == 'ReactJS') {
    questionText = 'What is the extension used for TypeScript in ReactJS?';
    options = ['.tsx', '.jsx', '.js', '.java'];
  } else if (selectedValue == 'React Native') {
    questionText =
      'What are the simpler of the two CLIs used to create React Native Apps?';
    options = ['Expo CLI', 'React Native CLI'];
  } else if (selectedValue == 'JavaScript') {
    questionText = 'Inside which HTML element do we put the JavaScript?';
    options = ['<javascript>', '<script>', '<js>', '<scripting>'];
  } else if (selectedValue == 'Java') {
    questionText =
      'In Java, the class name should always be different from file name. True or False?';
    options = ['True', 'False'];
  }
  const handleOptionPress = (index: number) => {
    setIndex(index);
  };
  const handleAnswerCheck = () => {
    setSubmit(true);
    if (selectedValue == 'ReactJS') {
      if (selectedIndex === 0 && score < maxScore) {
        setScore(score + 1);
      } else {
        setScore(currentScore);
      }
    } else if (selectedValue == 'React Native' && score < maxScore) {
      if (selectedIndex === 0) {
        setScore(score + 1);
      } else {
        setScore(currentScore);
      }
    } else if (selectedValue == 'JavaScript' && score < maxScore) {
      if (selectedIndex === 1) {
        setScore(score + 1);
      } else {
        setScore(currentScore);
      }
    } else if (selectedValue == 'Java' && score < maxScore) {
      if (selectedIndex === 1) {
        setScore(score + 1);
      } else {
        setScore(currentScore);
      }
    }
  };

  return (
    <View style={styles.container} testID="questionContainer">
      <View style={styles.questionNavigation}>
        {[
          'Question 1',
          'Question 2',
          'Question 3',
          'Question 4',
          'Question 5',
        ].map((question, index) => (
          <TouchableOpacity
            key={index}
            onPress={() => navigation.navigate(question)}
            style={styles.navigationTouchable}
            testID={`navigate-to-${index}`}>
            <Text style={styles.navigationText}>{index + 1}</Text>
          </TouchableOpacity>
        ))}
      </View>
      {trackScore && <Text testID="score">Score: {score}</Text>}
      <Text style={{fontSize: 20, color: 'black'}} testID="questionID">
        {questionText}
      </Text>
      {options.map((option, index) => (
        <Pressable
          key={index}
          style={[
            styles.optionStyling,
            selectedIndex === index && styles.selectedOption,
          ]}
          onPress={() => handleOptionPress(index)}>
          <Text testID={`answer-${index}`}>{option.toString()}</Text>
        </Pressable>
      ))}
      <Button
        title="Answer"
        onPress={handleAnswerCheck}
        disabled={submitted}
        testID="answerButton"
      />
      {selectedIndex !== null && (
        <Button
          testID="question2"
          title="Next Question"
          onPress={() => navigation.navigate('Question 2')}
        />
      )}
    </View>
  );
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  optionStyling: {
    backgroundColor: '#00A1E4',
    padding: 10,
    borderRadius: 20,
    marginTop: 6,
    marginBottom: 10,
    width: '70%',
  },
  selectedOption: {
    backgroundColor: '#CC3F0C',
  },
  questionNavigation: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    position: 'absolute',
    top: 0,
  },
  navigationTouchable: {
    backgroundColor: '#FEB95F',
    padding: 3,
    marginHorizontal: 7,
    borderWidth: 1,
    borderColor: 'black',
    borderRadius: 100,
    height: 30,
    width: 30,
    marginVertical: 20,
  },
  navigationText: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginLeft: 3,
  },
});
