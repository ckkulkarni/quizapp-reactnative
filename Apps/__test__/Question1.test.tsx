import Question1 from '../Question1';
import Adapter from 'enzyme-adapter-react-16';
import * as enzyme from 'enzyme';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack/lib/typescript/src/types';
import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import {QuizContext} from './../QuizContext';
enzyme.configure({adapter: new Adapter()});
const navigation = useNavigation<NativeStackNavigationProp<any>>();

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: jest.fn().mockReturnValue({
      navigate: jest.fn(),
    }),
  };
});

describe('Question 1 component', () => {
  const mockQuizContext = {
    score: 0,
    setScore: jest.fn(),
    name: 'chandu',
    setName: jest.fn(),
    email: 'chandra@chandu.com',
    setMail: jest.fn(),
    phone: '1234567890',
    setPhone: jest.fn(),
    selectedValue: 'ReactJS',
    setSelected: jest.fn(),
    trackScore: true,
    setTracker: jest.fn(),
  };

  it("the question is correct assuming that selectedValue state is equal to 'ReactJS'", () => {
    const screen = render(
      <QuizContext.Provider value={mockQuizContext}>
        <Question1 />
      </QuizContext.Provider>,
    );
    const questionText = screen.getByTestId('questionID');
    expect(questionText.props.children).toBe(
      'What is the extension used for TypeScript in ReactJS?',
    );
  });
  it('the answers are rendered correctly, assuming that the selected state value is ReactJS', async () => {
    await waitFor(() => {
      const screen = render(
        <QuizContext.Provider value={mockQuizContext}>
          <Question1 />
        </QuizContext.Provider>,
      );
      const answer0 = screen.getByTestId('answer-0');
      const answer1 = screen.getByTestId('answer-1');
      const answer2 = screen.getByTestId('answer-2');
      const answer3 = screen.getByTestId('answer-3');
      expect(answer0.props.children).toBe('.tsx');
      expect(answer1.props.children).toBe('.jsx');
      expect(answer2.props.children).toBe('.js');
      expect(answer3.props.children).toBe('.java');
    });
  });
  it('should update the score when the right answer is pressed', async () => {
    await waitFor(() => {
      const screen = render(
        <QuizContext.Provider value={mockQuizContext}>
          <Question1 />
        </QuizContext.Provider>,
      );
      fireEvent.press(screen.getByTestId('answer-0'));
      fireEvent.press(screen.getByTestId('answerButton'));
      expect(mockQuizContext.setScore).toBeCalledWith(1);
    });
  });
  it('score should remain the same when the wrong answer is pressed', async () => {
    await waitFor(() => {
      const screen = render(
        <QuizContext.Provider value={mockQuizContext}>
          <Question1 />
        </QuizContext.Provider>,
      );
      fireEvent.press(screen.getByTestId('answer-1'));
      fireEvent.press(screen.getByTestId('answerButton'));
      expect(screen.getByTestId('score').props.children[1]).toBe(0);
    });
  });
  it('when the next question button is pressed, the user should navigate to the next screen', async () => {
    await waitFor(() => {
      const screen = render(
        <QuizContext.Provider value={mockQuizContext}>
          <Question1 />
        </QuizContext.Provider>,
      );
      fireEvent.press(screen.getByTestId('question2'));
      expect(navigation.navigate).toBeCalledWith('Question 2');
    });
  });
  it('when one of the numbers at the top are pressed, it should navigate to the pressed question', async () => {
    await waitFor(() => {
      const screen = render(
        <QuizContext.Provider value={mockQuizContext}>
          <Question1 />
        </QuizContext.Provider>,
      );
      fireEvent.press(screen.getByTestId('navigate-to-1'));
      expect(navigation.navigate).toBeCalledWith('Question 2');
      fireEvent.press(screen.getByTestId('navigate-to-2'));
      expect(navigation.navigate).toBeCalledWith('Question 3');
      fireEvent.press(screen.getByTestId('navigate-to-3'));
      expect(navigation.navigate).toBeCalledWith('Question 4');
      fireEvent.press(screen.getByTestId('navigate-to-4'));
      expect(navigation.navigate).toBeCalledWith('Question 5');
    });
  });
});
