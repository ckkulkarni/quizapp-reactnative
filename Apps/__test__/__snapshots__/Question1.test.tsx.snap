// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question 1 component component renders properly 1`] = `ShallowWrapper {}`;
